Core/Base/FileManager.php

```diff
@@ -0,0 +1,347 @@
+     *                          by default SCANDIR_SORT_ASCENDING.
+     * @param array  $exclude   Array list of items to exclude, by default ['.', '..'].
+     * @param bool   $recursive Look for recursively or not, by default false.
+     *
+     * @return array
+     */
+    public function getFilesFrom($dir, $order = SCANDIR_SORT_ASCENDING, array $exclude = ['.', '..'], $recursive = false): array
+    {
+        $items = $this->getFrom($dir, $order, $exclude);
+
+        $moreItems = [];
+        foreach ($items as $pos => $item) {
+            if (is_dir($dir . self::DS . $item)) {
+                unset($items[$pos]);
+                if ($recursive) {
+                    foreach ($this->getFilesFrom($dir . self::DS . $item, $order, $exclude, $recursive) as $file) {
+                        $moreItems[] = str_replace(\FS_FOLDER . '/', '', $file);
+                    }
+                }
+            }
+        }
+
+
+        return array_merge($items, $moreItems);
+    }
+
+    /**
+     * Return a list of folders and files.
+     * Note: The path are relative to FS_FOLDER.
+     *
+     * @param array $directories Folder to start looking for files.
+     * @param int   $order       Order to apply SCANDIR_SORT_ASCENDING/SCANDIR_SORT_DESCENDING,
+     *                           by default SCANDIR_SORT_ASCENDING.
+     * @param array $exclude     Array list of items to exclude, by default ['.', '..'].
+     * @param bool  $recursive   Look for recursively or not, by default true.
+     *
+     * @return array
+     */
+    public function getAllFrom(array $directories, $order = SCANDIR_SORT_ASCENDING, array $exclude = ['.', '..'], $recursive = true): array
+    {
+        foreach ($directories as $directory) {
+            $items = $this->getFrom($directory, $order, $exclude);
+
+            $moreItems = [];
+            foreach ($items as $item) {
+                if ($recursive && is_dir($item)) {
+                    $moreItems[] = $item;
+                    foreach ($this->getAllFrom([$item], $order, $exclude, $recursive) as $file) {
+                        $moreItems[] = str_replace(\FS_FOLDER . '/', '', $file);
+                    }
+                } else {
+                    $moreItems[] = str_replace(\FS_FOLDER . '/', '', $item);
+                }
+            }
+            $result = array_unique(\array_merge($items, $moreItems));
+        }
+
+        sort($result);
+        return $result;
+    }
+
+    /**
+     * Create a folder and return the result.
+     * Returns True if success or False if fails.
+     *
+     * @param string $dir Folder to start looking for files.
+     * @param int    $perms
+     * @param bool   $recursive
+     *
+     * @return bool
+     */
+    public function createFolder($dir, $perms = self::PERMS_FOLDER, $recursive = true): bool
+    {
+        return !(!file_exists($dir) && !@mkdir($dir, $perms, $recursive) && !is_dir($dir));
```

Comment:
return !(!file_exists($dir) && !createFolder($dir, $perms, $recursive) && !is_dir($dir));

With this you can get infinite checks, to be sure it has not failed.

Reply:
> return !(!file_exists($dir) && !@mkdir($dir, $perms, $recursive) && !is_dir($dir));
Much simpler:
if (file_exists($dir) && @mkdir($dir, $perms, $recursive) && is_dir($dir)) {
    return true;
}
return false;

Reply:
I share this on the past https://github.com/kalessil/phpinspectionsea/blob/master/docs/probable-bugs.md#mkdir-race-condition

## 'mkdir(...)' race condition

This issue is difficult to reproduce, as any concurrency-related issues are.
It appears when several processes are attempting to create a directory which does not
yet exist. Specifically, when one process is between `is_dir()` and `mkdir()` after
another process has already managed to create the directory.

Apart from different code constructs which are vulnerable to this type of issue, the safe 
variant is following: `!is_dir($folder) && !mkdir($folder) && !is_dir($folder)`.